hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/Server.java

```diff
@@ -2879,9 +2881,12 @@ private void processRpcRequest(RpcRequestHeaderProto header,
             stateId = alignmentContext.receiveRequestState(
                 header, getMaxIdleTime());
             call.setClientStateId(stateId);
-            if (header.hasRouterFederatedState()) {
-              call.setFederatedNamespaceState(header.getRouterFederatedState());
-            }
+          }
+          if (header.hasRouterFederatedState()) {
+            call.setFederatedNamespaceState(header.getRouterFederatedState());
+          } else if (header.hasStateId()) {
+            // Only used to determine whether to return federatedNamespaceState.
+            call.setFederatedNamespaceState(EMPTY_BYTE_STRING);
```

Comment:
Can you expand on why we need this else part?

Reply:
@simbadzina Master, thanks for your review.

Here set the EMPTY_BYTE_STRING to identify that the client want to get all nameservice's stateId. So that `RouterStateIdContext` can set ResponseHeaderState.

I think there are too cases that identifying the client want to get the stateId from server.
1. header.hasStateId()
2. header.hasRouterFederatedState()